File: delivery/ios/demo/BlueSDK/DeviceStorage.swift
// DeviceStorage.swift
// BlueSDK Example - Bound device list local persistence
// Uses UserDefaults + Codable to store the bound device list

import Foundation

/// Bound device data model
struct BoundDevice: Codable {
    let deviceId: String           // UUID string (unique identifier)
    let deviceName: String         // Device broadcast name
    let bindTime: TimeInterval     // Bind timestamp
    var lastConnectedTime: TimeInterval // Last connected timestamp
}

/// Bound device local storage manager
final class DeviceStorage {

    static let shared = DeviceStorage()

    private let key = "bound_devices_v1"
    private let defaults = UserDefaults.standard

    private init() {}

    /// Load all bound devices (sorted by last connected time descending)
    func loadAll() -> [BoundDevice] {
        guard let data = defaults.data(forKey: key),
              let devices = try? JSONDecoder().decode([BoundDevice].self, from: data) else {
            return []
        }
        return devices.sorted { $0.lastConnectedTime > $1.lastConnectedTime }
    }

    /// Add a bound device (updates name if already exists)
    func add(_ device: BoundDevice) {
        var all = loadAll()
        if let idx = all.firstIndex(where: { $0.deviceId == device.deviceId }) {
            all[idx] = device
        } else {
            all.append(device)
        }
        persist(all)
    }

    /// Remove a bound device
    func remove(deviceId: String) {
        var all = loadAll()
        all.removeAll { $0.deviceId == deviceId }
        persist(all)
    }

    /// Update last connected time
    func updateLastConnected(deviceId: String) {
        var all = loadAll()
        if let idx = all.firstIndex(where: { $0.deviceId == deviceId }) {
            all[idx].lastConnectedTime = Date().timeIntervalSince1970
            persist(all)
        }
    }

    /// Check if device is bound
    func isBound(deviceId: String) -> Bool {
        return loadAll().contains { $0.deviceId == deviceId }
    }

    /// Clear all bindings
    func clearAll() {
        persist([])
    }

    // MARK: - Private

    private func persist(_ devices: [BoundDevice]) {
        if let data = try? JSONEncoder().encode(devices) {
            defaults.set(data, forKey: key)
        }
    }
}

File: delivery/android/demo/app/src/main/kotlin/com/blue/demo/DeviceStorage.kt
// DeviceStorage.kt
// BlueSDK Demo - Local persistence for bound device list
// Uses SharedPreferences + JSON to store bound devices

package com.blue.demo

import android.content.Context
import org.json.JSONArray
import org.json.JSONObject

/**
 * Bound device data model
 */
data class BoundDevice(
    val deviceId: String,           // MAC address (unique identifier)
    val deviceName: String,         // Device broadcast name
    val bindTime: Long,             // Bind timestamp
    val lastConnectedTime: Long = 0 // Last connection timestamp
)

/**
 * Bound device local storage manager
 * Uses SharedPreferences + JSON to persist bound device list
 */
object DeviceStorage {

    private const val PREFS_NAME = "bound_devices_prefs"
    private const val KEY_DEVICES = "bound_devices_v1"

    private fun prefs(context: Context) =
        context.getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE)

    /** Load all bound devices (sorted by last connected time, descending) */
    fun loadAll(context: Context): List<BoundDevice> {
        val json = prefs(context).getString(KEY_DEVICES, null) ?: return emptyList()
        return try {
            val arr = JSONArray(json)
            (0 until arr.length()).map { i ->
                val obj = arr.getJSONObject(i)
                BoundDevice(
                    deviceId = obj.getString("deviceId"),
                    deviceName = obj.getString("deviceName"),
                    bindTime = obj.getLong("bindTime"),
                    lastConnectedTime = obj.optLong("lastConnectedTime", 0)
                )
            }.sortedByDescending { it.lastConnectedTime }
        } catch (e: Exception) {
            emptyList()
        }
    }

    /** Add bound device (updates name if already exists) */
    fun add(context: Context, device: BoundDevice) {
        val all = loadAll(context).toMutableList()
        val existing = all.indexOfFirst { it.deviceId == device.deviceId }
        if (existing >= 0) {
            all[existing] = device
        } else {
            all.add(device)
        }
        persist(context, all)
    }

    /** Remove bound device */
    fun remove(context: Context, deviceId: String) {
        val all = loadAll(context).toMutableList()
        all.removeAll { it.deviceId == deviceId }
        persist(context, all)
    }

    /** Update last connected time */
    fun updateLastConnected(context: Context, deviceId: String) {
        val all = loadAll(context).toMutableList()
        val idx = all.indexOfFirst { it.deviceId == deviceId }
        if (idx >= 0) {
            all[idx] = all[idx].copy(lastConnectedTime = System.currentTimeMillis())
            persist(context, all)
        }
    }

    /** Check if device is already bound */
    fun isBound(context: Context, deviceId: String): Boolean {
        return loadAll(context).any { it.deviceId == deviceId }
    }

    /** Clear all bindings */
    fun clearAll(context: Context) {
        persist(context, emptyList())
    }

    // MARK: - Private

    private fun persist(context: Context, devices: List<BoundDevice>) {
        val arr = JSONArray()
        for (d in devices) {
            arr.put(JSONObject().apply {
                put("deviceId", d.deviceId)
                put("deviceName", d.deviceName)
                put("bindTime", d.bindTime)
                put("lastConnectedTime", d.lastConnectedTime)
            })
        }
        prefs(context).edit().putString(KEY_DEVICES, arr.toString()).apply()
    }
}
